ClickHouse.Client/Types/Grammar/Parser.cs

```diff
@@ -8,25 +8,60 @@ public static class Parser
         public static SyntaxTreeNode Parse(string input)
         {
             var tokens = Tokenizer.GetTokens(input).ToList();
+            
             var stack = new Stack<SyntaxTreeNode>();
             SyntaxTreeNode current = null;
-
+            uint level = 0;
             foreach (var token in tokens)
             {
+                // TODO: clean token from back-ticks?!
+                bool nestedChild = false;
                 switch (token)
                 {
                     case "(":
                         stack.Push(current);
+                        level++;
                         break;
                     case ",":
                         stack.Peek().ChildNodes.Add(current);
                         break;
                     case ")":
-                        stack.Peek().ChildNodes.Add(current);
+                        var peek = stack.Peek();
+                        if (!peek.Virtual)
+                        {
+                            peek.ChildNodes.Add(current);
+                        }
                         current = stack.Pop();
+                        level--;
+                        if (current.Virtual)
+                        {
+                            current = stack.Pop();
+                            level--;
+                        }
                         break;
                     default:
-                        current = new SyntaxTreeNode { Value = token };
+                        if ((token == "Nested" || token.EndsWith(" Nested")) && level >= 1)
```

Comment:
Clickhouse returns nested types in different ways depending on the level it's defined;
Level 0: Array(<type>) Ex; Array(String)
Level 1: Array(Nested(<fields>) Ex; Array(Nested(param_id UInt8, param_val String))
Level 2: <field name> Nested(<fields>) Ex; Array(Nested(param_id UInt8, param_val String, **sub_sub_nested Nested(param_id UInt8, param_val String)**))

I wouldn't say the parser knows anything about the types, but the way the string tokens can occur for a specific type.
Probably if you don't want to have any knowledge about the string to parse, this needs to be refactored to full-blown AST, to cope with the difference of definition. This seems to be a bit overkill.

What do you propose?

Reply:
What feature flag would you propose? This functionality also should only fail when you actually use nested sub-columns in the data set, which you wouldn't if you don't have a compatible clickhouse server.

Reply:
> I wouldn't say the parser knows anything about the types, but the way the string tokens can occur for a specific type.

You have specific behavior there for specific type which does not occur for other types

> Probably if you don't want to have any knowledge about the string to parse, this needs to be refactored to full-blown AST, to cope with the difference of definition. This seems to be a bit overkill.

Not sure why it needs to be "refactored into full-blown AST"? Right now, parsing `Nested(Nested)` works perfectly:
```
TypeConverter.ParseClickHouseType("Nested (UInt8, String, Nested (UInt8, String,  Nested(UInt8, String)))")
```

If the parser works for this, it works. If the problem is with serialization, it needs to be fixed in serialization.

There are 3 things at play here: 1.Parsing type string into tokens 2.creating a type out of these tokens 3.implementing some behavior based on type. If there is a type `Array(Nested(<fields>))` coming from ClickHouse, it must be represented as types `Array->Nested->[fields]` internally. Right now, you're:

- breaking roundtrip parsing/serialization: `TypeConverter.ParseClickHouseType(type).ToString() != type`
- losing information about parsed type

So basically, this PR introduces a breakage in a fundamental part of code to solve some very specific problem in another part of code

If you need some specific behavior when reading and writing Nested(Nested(...)) values, you need to change BinaryStreamReader/BinaryStreamWriter classes to achieve that

> What feature flag would you propose? This functionality also should only fail when you actually use nested sub-columns in the data set, which you wouldn't if you don't have a compatible clickhouse server.

See new `SupportedFeatures` property in `master` - previously it was a separate set of methods `Supports***()`. If we have different behavior based on server version, it's better to reflect it somewhere in API

Reply:
> Not sure why it needs to be "refactored into full-blown AST"? Right now, parsing `Nested(Nested)` works perfectly:

Maybe the parsing works, but CH will thread this as Nested->Array->Tuple. So it add's an additional "virtual" array.

If i understand you correctly you prefer to have separate implementations of the Nested type in the binary reader/writer and evaluate this logic for every row/column in the data-set. Instead of feeding the binary ready/writer with the correct type structure and being only processed once up-front?

I've added the supported feature from your PR #80

Reply:
You are presuming that parsing types as-is is incorrect, but it is not. **Correct** type structure is the one which is on the label, not the one which is "internally". `DateTime64` is internally a `decimal`, `DateTime` is internally a `UInt64`, `FixedString(N)` is internally an array of `char`s. Would you think it's correct to return `decimal` when asked for a `DateTime`?

Reply:
Of course a `DateTime64` type should be returned that way. I think the misconception is that the data you actually inputting is in the correct format/type, see for example my test case; `new[] { new[] { new object[] { 4, "v4" }, new object[] { 5, "v5" }, new object[] { 6, "v6" } }, new[] { new object[] { 9, "v9" } } }`
While this type structure is not reflected by the type string as returned by CH.

You could also simply run my unittests against your master branch, or simply disable these to;
https://github.com/olafbuitelaar/ClickHouse.Client/blob/master/ClickHouse.Client/Types/Grammar/Parser.cs#L43 
 if ((token == "Nested" || token.EndsWith(" Nested")) && level >= 1) =>  if (**false &&** (token == "Nested" || token.EndsWith(" Nested")) && level >= 1)
and 
https://github.com/olafbuitelaar/ClickHouse.Client/blob/master/ClickHouse.Client/Types/NestedType.cs#L34 remove this line
for example with the table2 test;
``` CREATE TABLE IF NOT EXISTS test_sub_table_x(id UInt32, params Nested (param_id UInt8, param_val String, sub_nested Nested (param_id UInt8, param_val String))) ENGINE = Memory```

you'll see the parser fails at; https://github.com/DarkWanderer/ClickHouse.Client/blob/master/ClickHouse.Client/Types/TypeConverter.cs#L119

And if you do get past those field name parsing's, the write will fail, since there is an mismatch in the type definition vs. the actual received data.